### PR TITLE
Smart iterator

### PR DESCRIPTION
**Pull request by uwieder**
Currently the 3D histogram enforces the rate

### CODE REVIEW

File: platform/src/main/java/org/hillview/sketches/BasicColStats.java [hunk 1]
```diff
@@ -23,6 +23,7 @@
 /**
  * A class that scans a column and collects basic statistics: maximum, minimum,
  * number of non-empty rows and the moments of asDouble values.
+ * TODO: fix the sampling version of it, or perhaps not allow sampling at all, or always enforce sampling.
```
[the remark this answers is not in the document]
Comment (uwieder) [reply platform/src/main/java/org/hillview/sketches/BasicColStats.java]: Nothing. The TODO was left by mistake

File: platform/src/main/java/org/hillview/table/api/IMembershipSet.java [hunk 2]
```diff
@@ -71,7 +71,7 @@ default IMembershipSet filter(IntPredicate predicate) {
      * @param seed  Random seed.
      * @return      An iterator over the sampled data.
      */
-    default IRowIterator getIteratorOverSample(double rate, long seed) {
+    default IRowIterator getIteratorOverSample(double rate, long seed, boolean enforceRate) {
```
[the remark this answers is not in the document]
Comment (uwieder) [reply platform/src/main/java/org/hillview/table/api/IMembershipSet.java]: Not sure I understand what you mean. This is part of the IMemberhipSet interface.

File: platform/src/main/java/org/hillview/table/membership/SparseMembershipSet.java [hunk 3]
```diff
@@ -101,10 +101,10 @@ public IRowIterator getIterator() {
      * Returns an iterator that runs over the sampled data.
      * @param rate  Sampling rate.
      * @param seed  Random seed.
-     * @return      An iterator over the sampled data.
+     * @return      An iterator over the sampled data. Using a lower rate is always beneficial
```
[the remark this answers is not in the document]
Comment (uwieder) [reply platform/src/main/java/org/hillview/table/membership/SparseMembershipSet.java]: this is the function getIteratorOverSample. Not sure I understand the comment

File: platform/src/main/java/org/hillview/table/api/IMembershipSet.java [hunk 4]
```diff
@@ -71,11 +71,12 @@ default IMembershipSet filter(IntPredicate predicate) {
      * @param seed  Random seed.
      * @return      An iterator over the sampled data.
      */
-    default IRowIterator getIteratorOverSample(double rate, long seed) {
-        if (rate >= 1)
-            return this.getIterator();
-        return this.sample(rate, seed).getIterator();
-    }
+    //default ISampledRowIterator getIteratorOverSample(double rate, long seed, boolean enforceRate) {
```
Comment (mbudiu-vmw) [inline platform/src/main/java/org/hillview/table/api/IMembershipSet.java]: this should be deleted.

File: platform/src/main/java/org/hillview/table/api/ISampledRowIterator.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package org.hillview.table.api;
+
+/**
+ * An iterator over the rows that were sampled with some rate
```
Comment (mbudiu-vmw) [inline platform/src/main/java/org/hillview/table/api/ISampledRowIterator.java]: this comment is incomplete; it was sampled from a MembershipSet.
The rate represents the sampling rate from the parent MembershipSet.

File: platform/src/main/java/org/hillview/table/membership/FullMembershipSet.java [hunk 6]
```diff
@@ -93,13 +95,46 @@ private IMembershipSet sampleUtil(final Randomness randomGenerator, final int k)
...
+     */
+    private double computeRate(double rate) {
+        if (rate  <= FullMembershipSet.samplingThreshold)
+            return rate;
+        else
+            return 1;
     }
 
-    private static class FullSampledRowIterator implements IRowIterator {
+    private static class NoSampleRowIterator implements ISampledRowIterator {
```
Comment (mbudiu-vmw) [inline platform/src/main/java/org/hillview/table/membership/FullMembershipSet.java]: Why is this duplicated?
There should be only one copy of this class.

**Pull request merged**